etcdserver: add auto compaction interval option

jrcichra opened the pull request:
With enough key revisions per second, the constant 5 minute interval on revision compaction isn't fast enough to keep up with growth. Make the interval configurable, which will override the default behavior.

Also pass the configuration through to periodic compaction for finer control of when compaction happens based on time.

Fixes:
https://github.com/etcd-io/etcd/issues/18471

jrcichra commented:
My implementation here tries to support both `periodic` and `revision` but my original focus was solving the problem for just `revision`. It may not make as much sense for `periodic`.

jmhbnz commented:
/ok-to-test

jrcichra commented:
I'm passing all the tests now. @jmhbnz how should I proceed? I see there's a codecov report but there's a large section of code I didn't touch regarding failed auto compaction.

moficodes commented:
/assign

jrcichra commented:
Rebase is complete 🎉

jrcichra commented:
Hey @moficodes have you had a chance to look at this change?

jrcichra commented:
/retest

jrcichra commented:
Hey @jmhbnz / @moficodes I've rebased and all the tests are passing. Could I get a review please? Thanks!

jmhbnz reviewed (commented):
Hi @jrcichra - Thanks for your effort working on this feature, it seems reasonable to allow the existing constant top be configurable. To speed up getting this accepted I think it would be best to include an e2e test.  There are some compaction related e2e tests in `tests/e2e/watch_test.go` which may provide some inspiration on potential approaches.

Commit:
etcdserver: add auto compaction interval option

With enough key revisions per second, the constant 5 minute interval on
revision compaction isn't fast enough to keep up with growth. Make the
interval configurable, which will override the default behavior.

Also pass the configuration through to periodic compaction for finer
control of when compaction happens based on time.

Fixes:
https://github.com/etcd-io/etcd/issues/18471

Signed-off-by: Justin Cichra <jcichra@cloudflare.com>

jrcichra committed:
Merge branch 'etcd-io:main' into configurable-revision-interval
